lsp/ARCHITECTURE.md

```diff
@@ -0,0 +1,110 @@
+# Architechture
+
+## Language Service
+
+The following is a class diagram for the underlying Language Service.
+These are the components that allow us to provide a text document and get
+back the appropriate language support results (hover, completion, ...).
+
+```mermaid
+classDiagram
+
+    class DocumentAST{
+        <<interface>>
+        +ASTNode root
+    }
+
+    class JsonDocumentAST {
+
+    }
+
+    class YamlDocumentAST {
+
+    }
+
+    DocumentAST <|.. JsonDocumentAST : Implements
+    JsonDocumentAST <|.. YamlDocumentAST : Extends
+
+    class DocumentParserStrategy {
+        <<interface>>
+        +parse(String text) DocumentAST
+        +canParse(TextDocument td) boolean
+    }
+
+    DocumentParserStrategy "0..*" --* "1" DocumentParserContext
+
+    class DocumentParserContext {
+        <<interface>>
+        +parse(TextDocument td): Optional~DocumentAST~
+    }
+
+    DocumentParserContext --> DocumentAST : Resolves
+    %% ---
+
+    class LanguageServiceStrategy {
+        <<interface>>
+        +getLanguageService(): LanguageService
+        +isMatch(TextDocument td, DocumentAST ast) boolean
+    }
+
+    LanguageServiceStrategy "0..*" --* "1" LanguageServiceContext
+
+    class BuildspecServiceStrategy {
+
+    }
+
+    BuildspecServiceStrategy ..|> LanguageServiceStrategy : Implements
+
+    class LanguageServiceContext {
+        <<interface>>
+        +getLanguageService(TextDocument td, DocumentAST ast) Optional~LanguageService~
+    }
+
+    LanguageServiceContext --> LanguageService : Resolves
+    LanguageServiceContext ..> DocumentAST : Depends On
+    LanguageServiceContext ..> TextDocument : Depends On
+
+    %% ---
+
+    class LanguageService {
+        <<interface>>
+        +completion(TextDocument td, DocumentAST ast, Position position) Promise~CompletionItem[] | CompletionList~
+        +diagnostic(TextDocument td, DocumentAST ast) Promise~Diagnostic[]~
+        +hover(TextDocument td, DocumentAST ast, Position position) Promise~Diagnostic[]~
+    }
+
+    LanguageService ..> DocumentAST : Depends On
+    LanguageService ..> TextDocument : Depends On
+
+    class BuildspecService {
+
+    }
+
+    BuildspecService ..|> LanguageService : Implements
+
+    %% ---
+
+    %% This is provided by the VSCode node module
```

Comment:
While looking at the document, I was wondering which portions belong to what modules. It looks like mermaid doesn't support the concept of packages (which could contain entities like classes). What do you think about adding things like `note for TextDocument "Provided by package vscode-foo"` so that we see the comments in-diagram?

Another idea would be to style/color the classes along their different ownerships, but that could get cumbersome to manage.

Are any other interfaces/classes the domain of other packages, and could those be annotated as well?